plugins/cryptolib/cryptolib.cpp

```diff
@@ -0,0 +1,635 @@
+     }
+     else if (strieq(algorithm, "aes-192-cbc"))
+     {
+         if (len != 24)
+             throw MakeStringException(-1, "Invalid Key Length %d specified for algorithm %s, try 24", len, algorithm);
+     }
+     else if (strieq(algorithm, "aes-256-cbc"))
+     {
+         if (len != 32)
+             throw MakeStringException(-1, "Invalid Key Length %d specified for algorithm %s, try 32", len, algorithm);
+     }
+     else
+         throw MakeStringException(-1, "Unsupported symmetric algorithm (%s) specified", algorithm);
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clSymmetricEncrypt(size32_t & __lenResult, void * & __result,
+                                                    const char * algorithm,
+                                                    const char * key,
+                                                    size32_t lenInputdata,const void * inputdata)
+{
+    verifySymmetricAlgorithm(algorithm, strlen(key) );
+
+    //Create a unique Initialization Vector
+    unsigned char iv[EVP_MAX_IV_LENGTH];
+    RAND_bytes(iv, EVP_MAX_IV_LENGTH);
+
+    //temporary buffer for result
+    MemoryBuffer out;
+    aesEncrypt(out, lenInputdata, inputdata, strlen(key), key, (const char *)iv);
+
+    //build result DATA buffer in the form:
+    // (size32_t)EVP_MAX_IV_LENGTH, IV, (size32_t)LenPlainText excluding NULL, (size32_t)LenCipher, Cipher
+    symmSerialize(__result, __lenResult, (const char *)iv, sizeof(iv), lenInputdata, out.length(), (const void *)out.bufferBase());
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clSymmetricDecrypt(size32_t & __lenResult,void * & __result,
+                                                const char * algorithm,
+                                                const char * key,
+                                                size32_t lenEncrypteddata,const void * encrypteddata)
+{
+    verifySymmetricAlgorithm(algorithm, strlen(key));
+
+    //Decompose DATA buffer
+
+    StringBuffer sbIV;
+    StringBuffer sbCipher;
+    size32_t     lenPlainText;
+    symmDeserialize(encrypteddata, sbIV, sbCipher, &lenPlainText);
+
+    __result = (char *)CTXMALLOC(parentCtx, lenPlainText);
+    __lenResult = lenPlainText;
+
+    MemoryBuffer decrypted;
+    size32_t len = aesDecrypt(decrypted, sbCipher.length(), sbCipher.str(), strlen(key), key, sbIV.str());
+    memcpy(__result, decrypted.toByteArray(), __lenResult);
+}
+
+
+
+//-------------------------------------------------------------------------------------------------------------------------------------------
+// HASH SUPPORT
+//-------------------------------------------------------------------------------------------------------------------------------------------
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clInstalledHashAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    loadDigests();
```

Comment:
Slightly cleaner to make the implementation of loadDigests() conditional on the version number - reduces the number of #ifs